hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/internal/EntityDelayedFetchInitializer.java

```diff
@@ -109,7 +116,28 @@ public void resolveInstance(RowProcessingState rowProcessingState) {
 		}
 		else {
 			final SharedSessionContractImplementor session = rowProcessingState.getSession();
-			final EntityPersister concreteDescriptor = referencedModelPart.getEntityMappingType().getEntityPersister();
+
+			final EntityPersister entityPersister = referencedModelPart.getEntityMappingType().getEntityPersister();
+			final EntityPersister concreteDescriptor;
+			if ( discriminatorAssembler != null ) {
+				concreteDescriptor = determineConcreteEntityDescriptor(
+						rowProcessingState,
+						discriminatorAssembler,
+						entityPersister
+				);
+				if ( concreteDescriptor == entityPersister && discriminatorAssembler.extractRawValue( rowProcessingState ) == null ) {
```

Comment:
I agree, I was reluctant to change that because of how it's used in `AbstractEntityInitializer` and I've debated creating an ad-hoc implementation here, but didn't want to duplicate logic. I've noticed, though, there are null checks there and the whole test suite passes even when returning null, so I've gone ahead and simplified it like you suggested (and added an assertion in `determineConcreteEntityDescriptor`).